block-layer/codecs/dag-cbor.md

```diff
@@ -37,93 +39,103 @@ In DAG-CBOR, map keys must be strings, as defined by the [IPLD Data Model]. Othe
 
 ## Strictness
 
-DAG-CBOR requires that there exist a single, canonical way of encoding any given object, and that encoded forms contain no superfluous data that may be ignored or lost in a round-trip decode/encode.
+DAG-CBOR requires that there exist a single, canonical way of encoding any given set of data, and that encoded forms contain no superfluous data that may be ignored or lost in a round-trip decode/encode.
 
 Therefore the DAG-CBOR codec must:
 
 1. Use no tags other than the CID tag (`42`). A valid DAG-CBOR encoder must not encode using any additional tags and a valid DAG-CBOR decoder must reject objects containing additional tags as invalid.
-   * This includes any of the initial values of the tag registry in [section 2.4 of the CBOR specification], such as dates, bignums, bigfloats, URIs, regular expressions and other complex, or simple values whether or not they map to the [IPLD Data Model].
-2. The only usable major type 7 minor types are those for encoding Floats (`25`, `26`, `27`), True (`20`), False (`21`) and Null (`22`).
-   * "Simple values" are not supported. This includes all registered or unregistered simple values that are encoded with a major type 7.
-   * Undefined (`23`) is not supported.
-3. Use the canonical CBOR encoding defined by the suggestions in [section 3.9 of the CBOR specification]. A valid DAG-CBOR decoder should reject objects not following these restrictions as invalid. Specifically:
+   * This includes any of the well defined tag numbers listed [section 3.4 of RFC 8949], such as dates, bignums, bigfloats, URIs, regular expressions and other complex, or simple values whether or not they map to the [IPLD Data Model].
+2. Use the "Deterministically Encoded CBOR" rule suggestions defined in [section 4.2 of RFC 8949] except for map key ordering, which follow the original rules as defined in [section 3.9 of RFC 7049]. Therefore, a valid DAG-CBOR encoder should produce encoded forms that adhere to the following rules, and a valid DAG-CBOR decoder should reject encoded forms not adhering to the following rules:
    * Integer encoding must be as short as possible.
    * The expression of lengths in major types 2 through 5 must be as short as possible.
    * The expression of tag numbers (specifically only `42`) must be as short as possible for major type 6. Therefore, for valid DAG-CBOR, the only tag token that can appear is `0xd82a` - where `0xd8` is "major type 6 with 8-bit integer to follow" and `0x2a` is the number `42`.
-   * The keys in every map must be sorted lowest value to highest. Sorting is performed on the bytes of the representation of the keys.
+   * The keys in every map must be sorted length-first by the byte representation of the string keys, where:
```

Comment:
I still hold that map sorting is an antifeature and it's almost certainly a mistake to specify it with a "must".

We've got lots of data in the wild that doesn't do this already, and lots of reasoning why it's ecosystemically dubious even to try to force it in the future.  I've written https://gist.github.com/warpfork/9315eac9848fa2c2ab36174272a80d9b about this and still feel it holds in its entirety.

I strongly suspect the outcomes of strict map sorting will be: people are pushed to use CBOR (and not DAG-CBOR) for handling any legacy data; and/or we'll see the emergence of a DAG-CBOR-HARDER codec in the future which is identical to this spec in every way but is order-preserving.  Or: we'll simply see almost every implementation not enforce this rule (in which case, what was the point?).

I would much rather we move this to a section on canonicalization which makes sorting *recommend* when the user data *is known to lack a meaningful ordering*.

I'm also not going to hold the PR up on this, because the previous text also stated sorting, and this text just states it better.

Reply:
OK, let me first acknowledge your overall objection here @warpfork, but this particular change to the text is only clarifying what the current status is, not changing the current status. It would be nice if we could find a way to split out this objection from this particular PR rather than hold up a clean-up with a structural objection. How can I adjust to help this happen? Do you want me to remove the change to "Final"?

I have thoughts on how we might inch toward a reconciliation on the views here, but I really don't want to bikeshed in this PR.

Reply:
Yeah, if we leave the final off for the moment being, I'd be happy overall.